stdlib/public/core/Algorithm.swift

```diff
@@ -75,6 +75,106 @@ public func max<T: Comparable>(_ x: T, _ y: T, _ z: T, _ rest: T...) -> T {
   return maxValue
 }
 
+/// Sorts the given arguments in ascending order, and returns the middle value.
+///
+///     // Values clamped to `0...100`
+///     median(0, .min, 100)  //-> 0
+///     median(0, .max, 100)  //-> 100
+///
+/// - Parameters:
+///   - x: A value to compare.
+///   - y: Another value to compare.
+///   - z: A third value to compare.
+///
+/// - Returns: The middle value.
+@_alwaysEmitIntoClient
+public func median<T: Comparable>(_ x: T, _ y: T, _ z: T) -> T {
+  var (x, y, z) = (x, y, z)
+  // Compare (and swap) each pair of adjacent variables.
+  if x > y {
+    (x, y) = (y, x)
+  }
+  if y > z {
+    (y, z) = (z, y)
+    if x > y {
+      (x, y) = (y, x)
+    }
+  }
+  // Now `x` has the least value, and `z` has the greatest value.
+  return y
+}
```

Comment:
I’m not an expert on optimization, but I put this into swift.godbolt.org alongside an alternative “sorting network” implementation that always performs 3 comparisons. To analyze them under optimized conditions, I made trivial wrappers that just use `Int` and compiled using “nightly” at “-O”.

The [godbolt link](https://swift.godbolt.org/z/3efxW5vEr) shows that the `Int`-specialized versions are nearly identical, except that your version (`intMedianA` in the link) performs one additional unconditional move at the start. They both have 3 compares and 4 conditional moves, the only difference is that the sorting-network version has only 1 move instead of 2.

Doing the same but with `Double` in place of `Int` shows an even larger difference, dropping the instruction count by 3. But I know you have a distinct version of `median` for `FloatingPoint`, so I’ll look at that separately.

• • •

The floating-point version with `isTotallyOrdered(belowOrEqualTo:)` [on godbolt](https://swift.godbolt.org/z/nsjf5P5Eh) is beyond my ability to evaluate.

Reply:
Thanks, those results are interesting, but not intuitive for me.
The only difference seems to be whether or not the last `if x > y` is nested.

--------------------------------------------------------------------------------

I'm not sure if the advice in [SE-0173][] on avoiding the [`swap(_:_:)`][] function is still valid?

> While `swap` was only intended to be used on collections, it is possible to use it on other variables. However, the recommended style for these uses is to not use a function at all

[SE-0173]:      <https://github.com/apple/swift-evolution/blob/main/proposals/0173-swap-indices.md>
[`swap(_:_:)`]: <https://github.com/apple/swift/blob/dac3e2b23e10fd06fcb740a7c4e6c752d4ef85f6/stdlib/public/core/MutableCollection.swift#L434-L461>

Reply:
> Thanks, those results are interesting, but not intuitive for me.
> The only difference seems to be whether or not the last `if x > y` is nested.

Right, it appears the compiler is rewriting things in whatever way it prefers.

One possibility is that the compiler prefers “branchless swaps”, so it transforms your version to avoid having two distinct execution paths, on the premise that compare-and-swap is faster than branching.

That transformation should make it equivalent to the sorting network version, but somehow the compiler doesn’t realize it can also eliminate the first move.

• • •

I didn’t know SE–0173 said that about `swap`. Replacing the `swap` with tuple assignment results in identical generated code.

Personally, I prefer the clarity of reading `swap` rather than having to mentally verify that the same variables appear on both sides on the assignment. Avoiding repetition and all.

But then again I would *also* prefer not to need the ampersands. Maybe for April fools I’ll propose a swap operator (“swoperator”?)…

```swift
infix operator <->

func <-> <T> (a: inout T, b: inout T) {
  swap(&a, &b)
}
```

Reply:
> To analyze them under optimized conditions, I made trivial wrappers that just use `Int` …

Alternatively, you can use [generic pre-specialization](https://github.com/apple/swift/blob/main/docs/Generics.rst#specialization), instead of the wrapper functions.

I also tried a version that used `min` and `max` — it was the same for `Int`, but worse for `Double` — and I don't know if a "stable sort" is possible.

```swift
@_specialize(exported: true, where T == Int)
@_specialize(exported: true, where T == Double)
public func median<T: Comparable>(_ x: T, _ y: T, _ z: T) -> T {
  max(
    min(x, y),
    min(
      max(x, y),
      z
    )
  )
}
```